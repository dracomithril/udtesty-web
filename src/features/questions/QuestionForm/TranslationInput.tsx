import styled from 'styled-components';
import React, { ChangeEvent } from 'react';
import { Input, Form } from 'antd';

export const TRANSLATION_INPUT_TEXTAREA = 'translation-input-textarea';
export const TRANSLATION_INPUT = 'translation_input';
const TranslationInput = styled(({
  label, name, value, onChange, className, disabled, needed, ...props
}: {
  label: string,
  name: string,
  disabled?: boolean;
  onChange?: (event: ChangeEvent<HTMLTextAreaElement>) => void
  value: string,
  needed?: boolean;
  'data-testid'?:string;
  className?: string;
  hidden?: boolean;
}) => {
  const hasError = !value && needed;

  return (
    <Form.Item
      label={label}
      required={needed}
      validateStatus={hasError ? 'error' : undefined}
      help={hasError ? 'nie powinno byc puste' : ''}
      className={className}
      data-testid={props['data-testid'] || TRANSLATION_INPUT}
    >
      <Input.TextArea
        disabled={disabled}
        data-testid={TRANSLATION_INPUT_TEXTAREA}
        name={name}
        onChange={onChange}
        value={value}
      />
    </Form.Item>
  );
}).attrs((props) => ({ fontWeight: props.needed ? 'bold' : 'unset' }))`
  display: ${(props) => (props.hidden ? 'none' : 'flex')};
  flex-direction: row;
  padding-top: 10px;
  justify-content: center;
  align-items: center;
`;

export default TranslationInput;
