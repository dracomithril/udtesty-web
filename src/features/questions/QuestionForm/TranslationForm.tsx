import React, { ChangeEvent } from 'react';
import { TranslationType, TranslationTypeKeys } from '@dracomithril/types';
import { LANGUAGES, LanguageType } from '../../../constants';
import TranslationInput from './TranslationInput';

export type TranslationFormChangeType = {
  name: string,
  value: TranslationType,
  isValid: boolean
};

export const TRANSLATION_FORM_TEST_ID = 'translation-form';

interface TranslationFormType {
  name: string,
  value: TranslationType;
  disabled?: boolean;
  'data-testid'?: string
  onChange?: (newValue: TranslationFormChangeType) => void,
  availableLanguages: TranslationTypeKeys[]
  type: 'choice' | 'question';
  className?: string;
}

const handleInputChange = (
  lang: TranslationTypeKeys,
  value: TranslationType,
  handler: (newValue: TranslationFormChangeType) => void,
) => ({ currentTarget }: ChangeEvent<HTMLTextAreaElement>) => {
  const translations = { ...value, [lang]: currentTarget.value };
  handler({
    name: currentTarget.name,
    value: translations,
    isValid: Object.values(translations).every((v) => !!v),
  });
};

const TranslationForm = ({
  value,
  type,
  name,
  onChange,
  className,
  disabled,
  availableLanguages = [],
  'data-testid': dataTestId,
}:
TranslationFormType) => {
  // todo hidden and needed?
  const inputs = (LANGUAGES as LanguageType[])
    .filter(({ code }) => availableLanguages.includes(code))
    .map(({ code, inPolish }: LanguageType) => {
      const id = `${[type, name, code].join('.')}`;
      const valueElement = value[code];
      return (
        valueElement != null && (
        <TranslationInput
          label={inPolish}
          data-testid={id}
          key={id}
          name={name}
          value={valueElement}
          needed={availableLanguages.includes(code)}
          hidden={!availableLanguages.includes(code)}
          disabled={disabled}
          onChange={onChange && handleInputChange(code, value, onChange)}
        />
        )
      );
    });
  return (
    <div data-testid={dataTestId} className={className}>
      {inputs}
    </div>
  );
};

TranslationForm.defaultProps = {
  className: undefined,
  disabled: false,
  'data-testid': TRANSLATION_FORM_TEST_ID,
  onChange: () => {},
};

export default TranslationForm;
