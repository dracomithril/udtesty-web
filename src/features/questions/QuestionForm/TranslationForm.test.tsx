import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import TranslationForm from './TranslationForm';
import { TRANSLATION_INPUT_TEXTAREA } from './TranslationInput';

describe('TranslationForm', () => {
  it('should render component', () => {
    const translation = { pl: '', ua: '' };
    const handleChange = jest.fn();
    const { baseElement, getByTestId } = render(<TranslationForm
      name="ABC"
      onChange={handleChange}
      value={translation}
      availableLanguages={['pl']}
      type="choice"
    />);
    expect(baseElement).toMatchSnapshot();
    const elementsByTagName = getByTestId(TRANSLATION_INPUT_TEXTAREA);
    const change = { target: { value: 'abcd' } };
    fireEvent.change(elementsByTagName, change);
    expect(handleChange)
      .toBeCalledWith(
        expect.objectContaining({
          value: expect
            .objectContaining({
              pl: change.target.value,
            }),
        }),
      );
  });
});
