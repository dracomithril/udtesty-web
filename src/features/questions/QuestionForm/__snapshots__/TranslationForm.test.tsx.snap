// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TranslationForm should render component 1`] = `
<body>
  <div>
    <div
      data-testid="translation-form"
    >
      <div
        class="ant-row ant-form-item ant-form-item-with-help sc-bdVaJa fCvIlw"
        data-testid="choice.ABC.pl"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class="ant-form-item-required"
            title="polski"
          >
            polski
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control-wrapper"
        >
          <div
            class="ant-form-item-control has-error"
          >
            <span
              class="ant-form-item-children"
            >
              <textarea
                class="ant-input"
                data-testid="translation-input-textarea"
                name="ABC"
              />
            </span>
            <div
              class="ant-form-explain show-help-appear"
            >
              nie powinno byc puste
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
